### PR TITLE
[mod] revision of the settings_loader

### LINKED ISSUE #3557: [Bug] New 'Hostnames plugin' not working with yml files
Vanilla 2024.6.7+f5eb56b6
Using official docker image

**What happened?**
Trying to change to new 'Hostnames plugin' as described here: https://github.com/searxng/searxng/discussions/3553)
Using a simple config with yml files, plugin fails to load due to an file not found error with the yml file

**How To Reproduce**
Switch to 'Hostnames plugin'
Use example config (and create the appropriate yml file)
Get error FileNotFoundError: File remove-hosts.yml does not exist!

**Expected behavior**
It works and loads the yml file

**Screenshots & Logs**

settings.yml
```shell
hostnames:
  replace: 'replace-hosts.yml'
  remove: 'remove-hosts.yml'
  high_priority:
  low_priority:

```

<details><summary>config files and permissions</summary>
<p>

docker exec -it searxng /bin/sh
/usr/local/searxng # ls -l /etc/searxng/
total 148
-rwxr-xr-x    1 10044    10044         1046 May 24 14:56 limiter.toml
-rwxr-xr-x    1 10044    10044       131003 Jun  7 14:05 remove-hosts.yml
-rwxr-xr-x    1 10044    10044          938 Jun  7 14:01 replace-hosts.yml
-rwxr-xr-x    1 10044    10044         7708 Jun  7 20:11 settings.yml
-rwxr-xr-x    1 10044    10044         1102 May  4 09:53 uwsgi.ini

</p>
</details> 

<details><summary>docker log</summary>
<pre>
2024-06-09 08:07:48,781 ERROR:searx.plugins: searx.plugins.hostnames: exception while loading, the plugin is disabled
Traceback (most recent call last):
  File "/usr/local/searxng/searx/plugins/__init__.py", line 115, in load_plugin
    plugin = import_module(plugin_module_name)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/usr/local/searxng/searx/plugins/hostnames.py", line 120, in <module>
    replacements = _load_regular_expressions_with_fallback('replace')
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/searxng/searx/plugins/hostnames.py", line 102, in _load_regular_expressions_with_fallback
    expressions = _load_regular_expressions(settings_key)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/searxng/searx/plugins/hostnames.py", line 88, in _load_regular_expressions
    setting_value = get_yaml_file(setting_value)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/searxng/searx/settings_loader.py", line 37, in get_yaml_file
    raise FileNotFoundError(f"File {file_name} does not exist!")
FileNotFoundError: File replace-hosts.yml does not exist!
2024-06-09 08:07:48,781 ERROR:searx.plugins: searx.plugins.hostnames: exception while loading, the plugin is disabled
Traceback (most recent call last):
  File "/usr/local/searxng/searx/plugins/__init__.py", line 115, in load_plugin
    plugin = import_module(plugin_module_name)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/usr/local/searxng/searx/plugins/hostnames.py", line 120, in <module>
    replacements = _load_regular_expressions_with_fallback('replace')
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/searxng/searx/plugins/hostnames.py", line 102, in _load_regular_expressions_with_fallback
    expressions = _load_regular_expressions(settings_key)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/searxng/searx/plugins/hostnames.py", line 88, in _load_regular_expressions
    setting_value = get_yaml_file(setting_value)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/searxng/searx/settings_loader.py", line 37, in get_yaml_file
    raise FileNotFoundError(f"File {file_name} does not exist!")
FileNotFoundError: File replace-hosts.yml does not exist!
2024-06-09 08:07:48,783 ERROR:searx.plugins: searx.plugins.hostnames: exception while loading, the plugin is disabled
Traceback (most recent call last):
  File "/usr/local/searxng/searx/plugins/__init__.py", line 115, in load_plugin
    plugin = import_module(plugin_module_name)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/usr/local/searxng/searx/plugins/hostnames.py", line 120, in <module>
    replacements = _load_regular_expressions_with_fallback('replace')
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/searxng/searx/plugins/hostnames.py", line 102, in _load_regular_expressions_with_fallback
    expressions = _load_regular_expressions(settings_key)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/searxng/searx/plugins/hostnames.py", line 88, in _load_regular_expressions
    setting_value = get_yaml_file(setting_value)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/searxng/searx/settings_loader.py", line 37, in get_yaml_file
    raise FileNotFoundError(f"File {file_name} does not exist!")
FileNotFoundError: File replace-hosts.yml does not exist!
2024-06-09 08:07:48,783 ERROR:searx.plugins: searx.plugins.hostnames: exception while loading, the plugin is disabled
Traceback (most recent call last):
  File "/usr/local/searxng/searx/plugins/__init__.py", line 115, in load_plugin
    plugin = import_module(plugin_module_name)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/usr/local/searxng/searx/plugins/hostnames.py", line 120, in <module>
    replacements = _load_regular_expressions_with_fallback('replace')
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/searxng/searx/plugins/hostnames.py", line 102, in _load_regular_expressions_with_fallback
    expressions = _load_regular_expressions(settings_key)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/searxng/searx/plugins/hostnames.py", line 88, in _load_regular_expressions
    setting_value = get_yaml_file(setting_value)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/searxng/searx/settings_loader.py", line 37, in get_yaml_file
    raise FileNotFoundError(f"File {file_name} does not exist!")
FileNotFoundError: File replace-hosts.yml does not exist!

</pre>
</details> 

**Additional context**
I triple checked that everything is as it should be, including the yml syntax, file permissions and location of the yml file that causes the FileNotFoundError.

### PR DESCRIPTION
**Pull request by return42**
The intention of this PR is to modernize the settings_loader implementations. The concept is old (remember, this is partly from 2014), back then we only had one config file, meanwhile we have had a folder with config files for a very long time.  Callers can now load a YAML configuration from this folder as follows ::

    settings_loader.get_yaml_cfg('my-config.yml')

- BTW this is a fix of #3557.

- Further the `existing_filename_or_none` construct dates back to times when there was not yet a `pathlib.Path` in all Python versions we supported in the past.

- Typehints & documentation have been added wherever appropriate

At the same time, this patch should also be downward compatible and not introduce a new environment variable. The localization of the folder with the configurations is further based on:

    SEARXNG_SETTINGS_PATH (wich defaults to /etc/searxng/settings.yml)

Which means, the default config folder is `/etc/searxng/`.

> [!IMPORTANT] 
> **intended functional changes!**
>
> If environment `SEARXNG_SETTINGS_PATH` was set and pointed to a not existing file, the previous implementation silently loaded the default configuration.  This behavior has been changed: if the file or folder does not exist, an EnvironmentError exception will be thrown in future.

Closes: https://github.com/searxng/searxng/issues/3557

### CODE REVIEW

File: searx/settings_loader.py [hunk 1]
```diff
@@ -1,68 +1,97 @@
...
-def load_yaml(file_name):
+def load_yaml(file_name: str | Path):
+    """Load YAML config from a file."""
     try:
         with open(file_name, 'r', encoding='utf-8') as settings_yaml:
-            return yaml.safe_load(settings_yaml)
+            cfg = yaml.safe_load(settings_yaml)
+            if cfg is None:
+                cfg = {}
+            return cfg
```
Comment (Bnyro) [inline searx/settings_loader.py]:
```suggestion
            return yaml.safe_load(settings_yaml) or {}
```

File: searx/settings_loader.py [hunk 2]
```diff
@@ -1,68 +1,97 @@
...
+def get_yaml_cfg(file_name: str | Path) -> dict:
+    """Shortcut to load a YAML config from a file, located in the
+
+    - :py:obj:`get_user_cfg_folder` or
+    - in the ``searx`` folder of the SearXNG instalation
+    """
 
+    folder = get_user_cfg_folder()
+    if not folder:
+        folder = Path(searx_dir)
```
Comment (Bnyro) [inline searx/settings_loader.py]:
```suggestion
    folder = get_user_cfg_folder() or Path(searx_dir)
```

File: searx/settings_loader.py [hunk 3]
```diff
@@ -1,68 +1,97 @@
...
-    # check the environment variable SEARXNG_SETTINGS_PATH
-    # if the environment variable is defined, this is the last check
-    if 'SEARXNG_SETTINGS_PATH' in environ:
-        return existing_filename_or_none(environ['SEARXNG_SETTINGS_PATH'])
+    folder = Path("/etc/searxng")
+    settings_path = os.environ.get("SEARXNG_SETTINGS_PATH")
+    if settings_path:
+        folder = Path(settings_path).parent
+        if not folder.is_dir():
+            raise EnvironmentError(1, f"Folder of {settings_path} not exists!", folder)
```
Comment (Bnyro) [inline searx/settings_loader.py]: Shouldn't we first check if
1. is `SEARXNG_SETTINGS_PATH` a folder (e.g. `/etc/searxng`)? if yes, that's our root settings path (the settings file is assumed to be named `settings.yml`)
2. we assume `SEARXNG_SETTINGS_PATH` is a file, thus we use its parent folder. if it exists, that's our settings path
3. we raise an exception
Currently, we only do step 2 and 3, but not step 1

File: searx/settings_loader.py [hunk 4]
```diff
@@ -1,68 +1,97 @@
...
 
-    # if SEARXNG_DISABLE_ETC_SETTINGS don't look any further
-    if environ.get('SEARXNG_DISABLE_ETC_SETTINGS', '').lower() in ('1', 'true'):
-        return None
+    elif os.environ.get('SEARXNG_DISABLE_ETC_SETTINGS', '').lower() in ('1', 'true'):
+        # This environment is intended exclusively for internal testing purposes
+        # and is therefore not documented!
+        folder = None
```
Comment (Bnyro) [inline searx/settings_loader.py]:
```suggestion
        return None
```

File: searx/settings_loader.py (hunk 3 above)
Comment (return42) [reply searx/settings_loader.py]: > Shouldn't we first check if … but not step 1

Good point I had already thought about it .. my first intention was not to change the meaning of `SEARXNG_SETTINGS_PATH`:

TL;DR;

Alternatively, I had thought about introducing a completely new environment variable for the config folder (SEARXNG_CFG_FOLDER).   Note that this settings loader is only for YAML files ... but we also have the toml file from the limiter .. and other plugins might bring more config file types ... all of them are in this folder.

But this would have the disadvantage that we have ambiguities (SEARXNG_CFG_FOLDER vs SEARXNG_SETTINGS_PATH).

Since I didn't want to choose one or the other with this PR, I decided not to change the meaning of SEARXNG_SETTINGS_PATH in this PR.

File: searx/settings_loader.py (hunk 4 above)
Comment (return42) [reply searx/settings_loader.py]: If we return here, we should also return two lines below .. None of the solutions is wrong or right, they all do what is intended. However, when implementing this function, I decided to return the value at the very end and not to exit the function within the conditions.

I'll leave it as it is for now ... or have I overlooked something?

Comment (Bnyro) [reply searx/settings_loader.py]: No, that's fine, it's just a matter of personal preference at this point, I understand your decision :)

File: searx/settings_loader.py (hunk 3 above)
Comment (Bnyro) [reply searx/settings_loader.py]: I agree that adding a new option may cause confusion and `SEARXNG_SETTINGS_PATH` should continue to be used.

I just think that it might be more convenient to also support folder paths, now that there's the ability to use files other than `settings.yml` (e.g. `replace.yml` for the `hostnames plugin`) it's really rather a folder path than a single file path. It wouldn't really change the behavior to implement 1. (the old full settings file path still works), it's just additional convenience for instance admins (in my opinion). Of course we will need to assume that the settings file is really named `settings.yml` in the case that a folder is provided.

File: tests/unit/test_settings_loader.py [hunk 5]
```diff
@@ -55,24 +51,22 @@ def test_is_use_default_settings(self):
...
-        with patch.dict(settings_loader.environ, {'SEARXNG_SETTINGS_PATH': '/dev/null'}):
-            settings, msg = settings_loader.load_settings()
-            self.assertTrue(msg.startswith('load the default settings from'))
-            self.assertEqual(settings['server']['secret_key'], "ultrasecretkey")
+        with patch.dict(os.environ, {'SEARXNG_SETTINGS_PATH': _settings("not_exists.yml")}):
+            with self.assertRaises(EnvironmentError):
+                _s, _m = settings_loader.load_settings()
+        with patch.dict(os.environ, {'SEARXNG_SETTINGS_PATH': "/folder/not/exists"}):
+            with self.assertRaises(EnvironmentError):
+                _s, _m = settings_loader.load_settings()
```
Comment (return42) [inline tests/unit/test_settings_loader.py:59]: > [!IMPORTANT] 
> **intended functional changes!**
>
> If environment `SEARXNG_SETTINGS_PATH` was set and pointed to a not existing file, the previous implementation silently loaded the default configuration.  This behavior has been changed: if the file or folder does not exist, an EnvironmentError exception will be thrown in future.

File: searx/settings_loader.py (hunk 3 above)
Comment (return42) [reply searx/settings_loader.py]: > I just think that it might be more convenient to also support folder paths,

Actually, you are right and my doubts are wrong ... I have now implemented it, see [diff](https://github.com/searxng/searxng/compare/c7d49e873b6a9ad787f0fcf59679d900934fa04f..1997d53c21bcd16ab70cd930aaa3a3c46394d674) .. further I changed the downward compatibility slightly; read my https://github.com/searxng/searxng/pull/3578#discussion_r1642433145

**Review by Bnyro — APPROVED** [review]
Thanks for applying my suggestions, everything looks good to me now :+1:

File: searx/settings_loader.py [hunk 6]
```diff
@@ -1,68 +1,114 @@
...
+            folder = settings_path.parent
+        else:
+            raise EnvironmentError(1, f"{settings_path} not exists!", settings_path)
 
-    # if SEARXNG_DISABLE_ETC_SETTINGS don't look any further
-    if environ.get('SEARXNG_DISABLE_ETC_SETTINGS', '').lower() in ('1', 'true'):
-        return None
+    if not folder and not disable_etc:
+        # default: rule 3.
+        folder = Path("/etc/searxng")
```
Comment (dalf) [inline searx/settings_loader.py:109]: The comment about the function says the code checks if `/etc/searxng` exists, and returns `None` otherwise.

File: tests/unit/test_settings_loader.py (hunk 5 above)
Comment (dalf) [reply tests/unit/test_settings_loader.py:59]: Why test `/folder/not/exists` in addition to `_settings("not_exists.yml")`?

File: searx/settings_loader.py (hunk 6 above)
Comment (return42) [reply searx/settings_loader.py:109]: Thanks for the hint :+1:  .. [its fixed now](https://github.com/searxng/searxng/compare/1997d53c21bcd16ab70cd930aaa3a3c46394d674..0f95b8b6ea1639e0b8df850edd0a4dfc7a2a54b8) .. I already had a hunch that I was missing something essential .. I've been looking at these code passages for too long not to make mistakes :-)

**Comment by dalf** [comment]
LGTM

File: tests/unit/test_settings_loader.py (hunk 5 above)
Comment (return42) [reply tests/unit/test_settings_loader.py:59]: Not really needed .. its just here in sense of testing "folder not exists" while the first test is in sense of "file not exists".  If you think its superfluous/redundant I can remove the second test.

Comment (dalf) [reply tests/unit/test_settings_loader.py:59]: It's more about the fact that the test fail if `/folder/not/exists` exists since this is folder outside the scope of the repository (I know the odds are really low). Feel free to ignore this comment.

Comment (return42) [reply tests/unit/test_settings_loader.py:59]: Yes, when the folder exists, the test will fail as expected:

```
$ sudo mkdir -p /folder/not/exists
$ make test.unit
...
======================================================================
FAIL: test_user_settings_not_found (tests.unit.test_settings_loader.TestUserSettings.test_user_settings_not_found)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/unit/test_settings_loader.py", line 58, in test_user_settings_not_found
    with self.assertRaises(EnvironmentError):
AssertionError: OSError not raised
```

> since this is folder outside the scope of the repository 

The probability that this absolute path name exists is low, so I would leave it as it is. Nevertheless, thank you for looking at this with a strict eye :+1:

**Review by dalf — APPROVED** [review]

**Comment by return42** [comment]
FYI: last force push was just a rebase on current master
